File: hariom_sanpra/public/py/bom.py
import frappe

@frappe.whitelist()
def set_bom(doc, method=None):
    if not doc.custom_link_to_main_product:
        return

    item = frappe.get_doc("Item", doc.custom_link_to_main_product)

    # Check if already exists
    already_exists = False
    for row in item.custom_link_bom:
        if row.bom == doc.name:
            already_exists = True
            break

    # Append only if not exists
    if not already_exists:
        item.append("custom_link_bom", {
            "bom": doc.name,
            "item_code": doc.item
        })
        item.save(ignore_permissions=True)

File: hariom_sanpra/public/js/bom.js
frappe.ui.form.on("BOM", {
	custom_total_rm_qty(frm) {
		(frm.doc.items || []).forEach((row) => set_bom_percent(frm, row));
		update_total_bom_percent(frm);
	},
});

frappe.ui.form.on("BOM Item", {
	qty(frm, cdt, cdn) {
		set_bom_percent(frm, locals[cdt][cdn]);
		update_total_bom_percent(frm);
	},
	custom_bom_percent_rm(frm) {
		update_total_bom_percent(frm);
	},
});

function set_bom_percent(frm, row) {
	const bom_qty = flt(frm.doc.custom_total_rm_qty);
	const percent = bom_qty ? (flt(row.qty) / bom_qty) * 100 : 0;

	frappe.model.set_value(
		row.doctype,
		row.name,
		"custom_bom_percent_rm",
		percent
	);
}
function update_total_bom_percent(frm) {
	let total = 0;

	(frm.doc.items || []).forEach((row) => {
		total += flt(row.custom_bom_percent_rm);
	});

	frm.set_value("custom_bom__rm_percent_", total);
}